### DISCUSSION TITLE
HTTP/2, já estão usando?

### DISCUSSION THREAD

**Discussion by felipefialho** [frontendbr/forum#103] [Discussão]
Fala pessoal, beleza?

Vi algumas palestras do @sergiolopes ano passado, na qual ele afirma que já estão usando na Caelum, e é totalmente tranquilo, já que existe uma especie de fallback para navegadores que não suportam, e esses vão ler o HTTP/1.

Estava conversando com o Artur Diniz também da Caelum  e ele disse que o novo site da [Alura](https://www.alura.com.br/) também já está usando HTTP/2.

Seria sensacional, especialmente por não ter necessidades de sprites (SVG Icons, olá) e poder ter uma componentização absurda de CSS e JavaScript, já que as boas praticas são inversas: Não devemos concatenar o código.

Penso inclusive em fazer uma modularização "nativa", deixando o browser carregar por default apenas CSS e JavaScript estruturais e cada componente injetar seu CSS e JavaScript quando necessário.

As possibilidades são muitas.

Enfim, já estão pensando em usar? Não acho que seria um problema, visto que navegadores antigos funcionaram normalmente, mas sem as vantagens do HTTP/2.

**Comment by rickbenetti** [comment]
O HTTP/2 é animal, porém o cara da infra aqui acha que é gambiarra ainda e não recomenda o uso, eu tava loco pra testar a dobradinha PHP7 + HTTP/2 pois vi que sites como o Smashing Magazine já estão usando: https://www.smashingmagazine.com/2016/02/getting-ready-for-http2/

**Comment by rubenmarcus** [comment]
a Digital Ocean ainda não oferece, mas pelo que vi dá pra implementar o CloudFlare em conjunto com ela, é isso mesmo?

https://www.digitalocean.com/community/questions/how-do-i-setup-cloudflare-and-digitalocean-with-my-droplet

https://www.youtube.com/watch?v=iWaFAIRWsA4 ??

**Comment by lagden** [comment]
eu uso! Configurei no Apache, mas já utilizava o SPDY antes...

**Comment by sergiolopes** [comment]
Gente, HTTP/2 é só sucesso! Vejo zero motivos pra não migrar tudo.
E estou num caso de amor com server push :heart:

**Comment by felipefialho** [comment]
@sergiolopes Como você tem trabalhado o CSS e o JavaScript? 

Eu estava pensando em criar arquivos base para a estrutura geral (padrão em todas as páginas) e só incluir o restante quando necessário em determinadas páginas e widgets (já que não vai mais ser preciso concatenar).

Acha uma ideia interessante ou acredita ter outra solução mais legal?

**Comment by sergiolopes** [comment]
To experimentando ainda tbm. Eu nao desconcatenaria tudo pq tem outros overheads que contam tbm. Aí fazer, sei la, 200 requests não fica tão bom. Mas tbm nao precisa ser 1 arquivão só. Tenho feito 1 base e 1 ou 2 por pagina/navegacao pro CSS. E server-pushando todos os CSS do above the fold.

JavaScript carrego tudo 100% assíncrono então ja deixo mais granular pq nao vai bloquear a renderizacao mesmo.

---

Tenho um artigo interessante da Khan Academy sobre desconcatenar no HTTP/2 onde eles acharam alguns gargalos com muitos arquivos: http://engineering.khanacademy.org/posts/js-packaging-http2.htm

**Comment by dougaraujos** [comment]
@sergiolopes Interessante a questão do **zlib**. Esse case da Khan Academy me fez repensar a implementação do protocolo.

**Comment by lagden** [comment]
Eu uso o https://wiki.mozilla.org/LZMA2_Compression junto com o Apache.
Dessa forma os arquivos de produção ficam todos `gz` e `xz`.
Então o script de `xxx.min.js` de `700K` fica `186K`.
Geralmente é legal para projetos `SPA`

**Comment by icaromh** [comment]
eu to louco pra usar também! Ótimo ver que já tão usando e boa.

@sergiolopes Vocês tão usando com nginx lá?

**Comment by sergiolopes** [comment]
Uma das Apps no App Engine e outra no nginx. Ambas com HTTP/2, bem sossegado de habilitar

**Comment by mandado** [comment]
@sergiolopes  você usa o defer para carregar ou usa o async ? como você controla a manipulação do dom, já que está carregando assíncrono ?

**Comment by sergiolopes** [comment]
Uso mais `async`. Maioria dos scripts no fim da pg, então o DOM ja esta pronto. Quando preciso de um script no topo (pra priorizar), manipulo os elementos no DOMContentLoad mesmo.

E para dependencias entre scripts assincronos costumo ir de AMD (quando preciso)

**Comment by eduardojmatos** [comment]
Trazendo a discussão de volta aqui, pergunta pro @sergiolopes. Vocês não usam AWS né? Como é que vocês estão servindo esses assets? Usam cloudflare - que já suporta Server Push?

**Comment by sergiolopes** [comment]
Usamos aws sim. A plataforma de cursos da alura está na Amazon, com nginx
(e aí sem Server push pq ele não suporta ainda).

A home da alura e o site da Caelum estão no Google App Engine que suporta
server push direto.

Em ambos os casos os assets são servidos pela própria aplicação, não usamos
CDN.

**Comment by eduardojmatos** [comment]
@sergiolopes Boa, saquei!

Escolheram Google App Engine por algum motivo específico? Tenho visto bastante gente usar Cloudflare pra esse esquema

**Comment by raphaelfruneaux** [comment]
@rubenmarcus a Digital Ocean já oferece. Olha: [Nginx - Ubuntu 16.04](https://www.digitalocean.com/community/tutorials/how-to-set-up-nginx-with-http-2-support-on-ubuntu-16-04)

**Comment by sergiolopes** [comment]
@eduardojmatos Usamos GAE há 7 anos, bem antes do HTTP/2 existir. Os motivos pra usá-lo são muitos (facilidade, custo, escalabilidade simples etc). O HTTP/2 acabou vindo depois como uma feliz consequência de estarmos lá já.

Se a questão é sobre como usar o HTTP/2 em produção, não vejo problemas em ir de nginx, apache etc ao invés de um PaaS tipo o GAE.

**Comment by eduardojmatos** [comment]
Massa! Bom saber =D

**Comment by RatoX** [comment]
Revivendo essa thread pq deu umas merdas ai.

Tá complicado de usar o HTTP 2 no chrome, depois que eles retiraram o [suporte para o NPN ](https://www.nginx.com/blog/supporting-http2-google-chrome-users/) ficamos meio que sem opções: Recompilar o NGINX do fonte ou migrar pra um Ubuntu 16.

Isso meio que me frustou no uso do HTTP 2, servidores parudos como o Debian 8 ficaram inúteis.


@LFeh O que tu fez com seus servidores?

**Comment by rickbenetti** [comment]
E passado um tempo agora uso AWS com HTTP2 e tá sendo normal, bem mais sussegado do que eu esperava.
@RatoX realmente não imagino o trampo que seja pensar numa possível migração pra outra distro ou encontrar formas de fazer funcionar, pois acabei mantendo tudo com Ubuntu 16 mesmo, como as configurações dos meus projetos são enxutas, isso facilita muito.

**Comment by RatoX** [comment]
@rkb81 pra projetos novos e que já rodam na distro do Ubuntu é bem de boa mas a galera que usa o Debian por examplo...
